build, qt: Fix cross-compiling detection on M1-based macOS (aarch64)

hebasto opened the pull request:
#20641 made building of the `qt` package in depends more structured, and, particularly, fixed cross-compiling detection on Intel-based macOS (x86_64).

However, a similar issue still present on M1-based macOS (arm64):

```
% make -C depends qt_configured
...
Configure summary:

Building on: macx-clang (x86_64, CPU features: cx16 mmx sse sse2 sse3 ssse3 sse4.1)
Building for: macx-clang (arm64, CPU features: neon crc32)
Target compiler: clang (Apple) 12.0.5
Configuration: cross_compile largefile neon precompile_header silent release c++11 c++14 c++1z reduce_exports static stl
...
```

Note an unexpected `cross_compile`.

```
% make -C depends qt_configured HOST=x86_64-apple-darwin18
...
Configure summary:

Build type: macx-clang (x86_64, CPU features: cx16 mmx sse sse2 sse3 ssse3 sse4.1)
Compiler: clang (Apple) 12.0.5
Configuration: sse2 aesni sse3 ssse3 sse4_1 sse4_2 avx avx2 avx512f avx512bw avx512cd avx512dq avx512er avx512ifma avx512pf avx512vbmi avx512vl f16c largefile precompile_header rdrnd shani silent x86SimdAlways release c++11 c++14 c++1z reduce_exports static stl
...
```

Note that the build system failed to detect cross compiling.

In both cases building the configured `qt` package fails without installed Rosetta 2 (see #22402).

This PR:
- fixes both cases above
- makes #22402 no longer required, and reverts it
- does not change the build system behavior on Intel-based macOS (x86_64)
- does not touch the code for cross-compiling on Linux builder, see a dedicated #21851

With this PR:
```
% make -j 9 -C depends qt_configured
...
Configure summary:

Build type: macx-clang (arm64, CPU features: neon crc32)
Compiler: clang (Apple) 12.0.5
Configuration: largefile neon precompile_header silent release c++11 c++14 c++1z reduce_exports static stl
...
```

```
% make -j 9 -C depends qt_configured HOST=x86_64-apple-darwin18
...
Configure summary:

Building on: macx-clang (arm64, CPU features: neon crc32)
Building for: macx-clang (x86_64, CPU features: cx16 mmx sse sse2 sse3 ssse3 sse4.1)
Target compiler: clang (Apple) 12.0.5
Configuration: cross_compile sse2 aesni sse3 ssse3 sse4_1 sse4_2 avx avx2 avx512f avx512bw avx512cd avx512dq avx512er avx512ifma avx512pf avx512vbmi avx512vl f16c largefile precompile_header rdrnd shani silent x86SimdAlways release c++11 c++14 c++1z reduce_exports static stl
...
```

This PR is based on #22506.

hebasto committed:
build, qt: Fix cross-compiling detection on M1-based macOS (aarch64)

hebasto committed:
Revert "doc: Install Rosetta on M1-macOS for qt in depends"

This reverts commit cdb41d5573b1e2ed1bc1d8d1dc9f77e82672ee1f.

It is no longer required because of the previous commit.

hebasto commented:
cc @promag @jarolrod @fanquake

jarolrod commented on depends/patches/qt/fix_darwin_archs.patch:
```diff
@@ -0,0 +1,16 @@
+To cross compile for the x86_64-darwin host on aarch64-darwin,
+passing QMAKE_APPLE_DEVICE_ARCHS=arm64 to the configure script
+does not work for some reasons.
```
I would prefer if we could document the reason it doesn't work. And any possible information on how/when/what circumstances this patch could be dropped

hebasto replied:
> I would prefer if we could document the reason it doesn't work.

I failed to track this bug down, unfortunately.

> And any possible information on how/when/what circumstances this patch could be dropped

As this is an ad-hoc hack, we should try to drop it while bumping Qt version in the future.

hebasto replied:
I think the #23583 is the correct solution for this problem.

jarolrod reviewed (commented):
Concept ACK, thanks for fixing this. One concern: https://github.com/bitcoin/bitcoin/pull/22487/files#r671973286

Tested on an M1 Machine. Confirming that with this change we can drop the recommendation to install rosseta.

**master:**

a) native depends:
```
Configure summary:

Building on: macx-clang (x86_64, CPU features: cx16 mmx sse sse2 sse3 ssse3 sse4.1)
Building for: macx-clang (arm64, CPU features: neon crc32)
Target compiler: clang (Apple) 12.0.5
Configuration: cross_compile largefile neon precompile_header silent release c++11 c++14 c++1z reduce_exports static stl
```

b) building for x64:
```
Configure summary:

Build type: macx-clang (x86_64, CPU features: cx16 mmx sse sse2 sse3 ssse3 sse4.1)
Compiler: clang (Apple) 12.0.5
Configuration: sse2 aesni sse3 ssse3 sse4_1 sse4_2 avx avx2 avx512f avx512bw avx512cd avx512dq avx512er avx512ifma avx512pf avx512vbmi avx512vl f16c largefile precompile_header rdrnd shani silent x86SimdAlways release c++11 c++14 c++1z reduce_exports static stl
```

**pr:**

a) native depends:

```
Configure summary:

Build type: macx-clang (arm64, CPU features: neon crc32)
Compiler: clang (Apple) 12.0.5
Configuration: largefile neon precompile_header silent release c++11 c++14 c++1z reduce_exports static stl
```

b) building for x64:
```
Configure summary:

Building on: macx-clang (arm64, CPU features: neon crc32)
Building for: macx-clang (x86_64, CPU features: cx16 mmx sse sse2 sse3 ssse3 sse4.1)
Target compiler: clang (Apple) 12.0.5
Configuration: cross_compile sse2 aesni sse3 ssse3 sse4_1 sse4_2 avx avx2 avx512f avx512bw avx512cd avx512dq avx512er avx512ifma avx512pf avx512vbmi avx512vl f16c largefile precompile_header rdrnd shani silent x86SimdAlways release c++11 c++14 c++1z reduce_exports static stl
```

While here

I noticed the following line in the output when building qt through depends, how much of a concern can this be?:

```
> main.cpp:5:12: fatal error: 'linux/input.h' file not found
> #  include <linux/input.h>
>            ^~~~~~~~~~~~~~~
> 1 error generated.
> make[1]: *** [main.o] Error 1
test config.qtbase_gui.tests.evdev FAILED
```

fanquake commented:
> I noticed the following line in the output when building qt through depends, how much of a concern can this be?:

promag reviewed (commented):
Tested ACK 5a55ba9b04f6e65e4e06564a092205c41d7de432.

However, I tried the following (hardcoded) and it worked fine:

```
$(package)_config_opts_darwin += -device-option QMAKE_APPLE_DEVICE_ARCHS=x86_64

make -C depends qt_configured HOST=x86_64-apple-darwin18
```

```
$(package)_config_opts_darwin += -device-option QMAKE_APPLE_DEVICE_ARCHS=arm64

make -C depends qt_configured
```

Note that it doesn't use `_config_opts_x86_64_darwin` or `_config_opts_aarch64_darwin`.

hebasto commented:
@promag 
> However, I tried the following (hardcoded) and it worked fine:

What is your Qt Configure summary?

fanquake reviewed (commented):
Does configuring with qt, for a x86-64 cross-compile, on an M1, using this PR, currently work?

fanquake commented on depends/patches/qt/fix_darwin_archs.patch:
```diff
@@ -0,0 +1,16 @@
+ 
+ QMAKE_MACOSX_DEPLOYMENT_TARGET = 10.12
+-QMAKE_APPLE_DEVICE_ARCHS = x86_64
++QMAKE_APPLE_DEVICE_ARCHS = $$system("uname -m")
```
Right, we can't do that because we only use this for the Linux cross-compile (`macx-clang-linux`). Kind of annoying that this needs even more hacky patching. Although I guess that's what you get when trying to force support for things like this into Qt versions that well and truly don't support it.

hebasto committed:
depends: Define -arch for compilers on darwin builder

hebasto commented:
Updated 5a55ba9b04f6e65e4e06564a092205c41d7de432 -> 8edac79a405b259544db12654e502e15717f18fc ([pr22487.01](https://github.com/hebasto/bitcoin/commits/pr22487.01) -> [pr22487.02](https://github.com/hebasto/bitcoin/commits/pr22487.02)):

- rebased on top of the #22506 to make possible to cross-compile `bitcoin-qt` binaries

hebasto commented:
@fanquake 
> Does configuring with qt, for a x86-64 cross-compile, on an M1, using this PR, currently work?

Yes, it does, after the [recent update](https://github.com/bitcoin/bitcoin/pull/22487#issuecomment-883483311).

hebasto commented:
Rebased 8edac79a405b259544db12654e502e15717f18fc -> d660e8057c756dd0262cacf86b97b2519749579e ([pr22487.02](https://github.com/hebasto/bitcoin/commits/pr22487.02) -> [pr22487.03](https://github.com/hebasto/bitcoin/commits/pr22487.03)) due to the conflict with #22526.

hebasto commented:
Rebased d660e8057c756dd0262cacf86b97b2519749579e -> 85888b5210c4c58cc82a06907d0d99ab1eed7f9c ([pr22487.03](https://github.com/hebasto/bitcoin/commits/pr22487.03) -> [pr22487.04](https://github.com/hebasto/bitcoin/commits/pr22487.04)) due to the conflict with #22448.

jarolrod commented:
GUIX hashes

```
find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum

df863a06c4b6b708580cb82e807194793b9c3a8fb9396a66c0099c0c29617d6b  guix-build-85888b5210c4/output/aarch64-linux-gnu/SHA256SUMS.part
6e43e1faf23c13460a779aff681c0b39ab489cdc7ffc08656492638f0d1e9745  guix-build-85888b5210c4/output/aarch64-linux-gnu/bitcoin-85888b5210c4-aarch64-linux-gnu-debug.tar.gz
a23d0c0734f80e6faa15da864a71670131ab4ca0842a866585137f253d4aa6d7  guix-build-85888b5210c4/output/aarch64-linux-gnu/bitcoin-85888b5210c4-aarch64-linux-gnu.tar.gz
3d8f004f946919d0970ab7a01b8fc8d2708f67297e2e106009428b0512c13787  guix-build-85888b5210c4/output/arm-linux-gnueabihf/SHA256SUMS.part
bd6bc1c40cf664d0044ba21b4d026ba4b43d4b14b70552fc0ddbf3a93689c519  guix-build-85888b5210c4/output/arm-linux-gnueabihf/bitcoin-85888b5210c4-arm-linux-gnueabihf-debug.tar.gz
b0cb8eede06b51a359d7d8fed7adf4b1397d814b68d2e423470df62ba1fcca83  guix-build-85888b5210c4/output/arm-linux-gnueabihf/bitcoin-85888b5210c4-arm-linux-gnueabihf.tar.gz
b41950cc73a0130989c30093a27982ebfabb9f8de31f9310f06fa6e14644eeea  guix-build-85888b5210c4/output/dist-archive/bitcoin-85888b5210c4.tar.gz
57d07de9fe2e43a1b5cc508e13a517f7d224085f19cb169627b97d4e9ba61f18  guix-build-85888b5210c4/output/powerpc64-linux-gnu/SHA256SUMS.part
07f69a38755900293550aab3c603ed46193e6f3737e0dce3419bbe7e9d96a226  guix-build-85888b5210c4/output/powerpc64-linux-gnu/bitcoin-85888b5210c4-powerpc64-linux-gnu-debug.tar.gz
9469b2c83c5df17717ddab52015a352256fcf3241bd500321438558ceec8d21c  guix-build-85888b5210c4/output/powerpc64-linux-gnu/bitcoin-85888b5210c4-powerpc64-linux-gnu.tar.gz
57c2a2384e4e0feeaf5c261e0ba05652253e10bb84c1597e8b18c0b280ed0d6d  guix-build-85888b5210c4/output/powerpc64le-linux-gnu/SHA256SUMS.part
a3587852be8dba91dac10a91f8827570677cae215afb61865dfa962b4d92044b  guix-build-85888b5210c4/output/powerpc64le-linux-gnu/bitcoin-85888b5210c4-powerpc64le-linux-gnu-debug.tar.gz
b7a2b4e450a76bae79da8012b1b9ef70467d3760d61bc8a0d8b41271249caf44  guix-build-85888b5210c4/output/powerpc64le-linux-gnu/bitcoin-85888b5210c4-powerpc64le-linux-gnu.tar.gz
cefaeb09ff6f6ed57850a1762090b58d118515d0e953d1d5eea1ad031f73dafa  guix-build-85888b5210c4/output/riscv64-linux-gnu/SHA256SUMS.part
b6897612ddca28534ffa61fd2bb321eafabcb01f3e16140934a732086930b851  guix-build-85888b5210c4/output/riscv64-linux-gnu/bitcoin-85888b5210c4-riscv64-linux-gnu-debug.tar.gz
90a4c28c2c85c562a0ba50cc158867d4634b6fa961fa61ec319fb87318f1780b  guix-build-85888b5210c4/output/riscv64-linux-gnu/bitcoin-85888b5210c4-riscv64-linux-gnu.tar.gz
ad48cd61446fbfc2e982aa2f6facced8b8c63c2606912fafea1eceadcf0ef1bd  guix-build-85888b5210c4/output/x86_64-apple-darwin18/SHA256SUMS.part
b3f866ed75e3e3eff5d6c565645abf7e6083de99c4c745d40c3053c60cebf7e5  guix-build-85888b5210c4/output/x86_64-apple-darwin18/bitcoin-85888b5210c4-osx-unsigned.dmg
c31f297161e002dd7e61bdc1ae511d6b39093c429cfe3b8f98ab51da3eca25eb  guix-build-85888b5210c4/output/x86_64-apple-darwin18/bitcoin-85888b5210c4-osx-unsigned.tar.gz
43835f47aeaa88c3b80b6623a9642f7ff6c887a41c30ab3c5f45ddafc004171b  guix-build-85888b5210c4/output/x86_64-apple-darwin18/bitcoin-85888b5210c4-osx64.tar.gz
f9a7c42bdf86aa1e16508f39f8144d3dd051a0a729fe3ffb6710e10723214ee5  guix-build-85888b5210c4/output/x86_64-linux-gnu/SHA256SUMS.part
da8b9e48cf9470e7105bd0614fefd618445cba9150a5ed3ca56b0a8f44133909  guix-build-85888b5210c4/output/x86_64-linux-gnu/bitcoin-85888b5210c4-x86_64-linux-gnu-debug.tar.gz
fbe71ec235394aed4643f5c48d4249a17b3769218757b5ac33c5a945d61c3236  guix-build-85888b5210c4/output/x86_64-linux-gnu/bitcoin-85888b5210c4-x86_64-linux-gnu.tar.gz
f10c3f187b9f19dcd0306e84bb81b06234bc48164377f9488ed457dfe024d438  guix-build-85888b5210c4/output/x86_64-w64-mingw32/SHA256SUMS.part
b59fed910a459911d7e088b5f9996ff2a5e271da1ccd3f20adc512fb64e73154  guix-build-85888b5210c4/output/x86_64-w64-mingw32/bitcoin-85888b5210c4-win-unsigned.tar.gz
b5314730b0cf1176821bbc66f9c9ecec4e97bb189733533a80f043e0db49167f  guix-build-85888b5210c4/output/x86_64-w64-mingw32/bitcoin-85888b5210c4-win64-debug.zip
472b673f781d1dd0f411ef8ee42baf40e01592c82ec8538ab7df70c9758318c5  guix-build-85888b5210c4/output/x86_64-w64-mingw32/bitcoin-85888b5210c4-win64-setup-unsigned.exe
185abcc544ea7c625be27a866e804b23885e61abbb28937ed75976996fd39668  guix-build-85888b5210c4/output/x86_64-w64-mingw32/bitcoin-85888b5210c4-win64.zip
```

hebasto commented:
#### Guix builds:

```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
df863a06c4b6b708580cb82e807194793b9c3a8fb9396a66c0099c0c29617d6b  guix-build-85888b5210c4/output/aarch64-linux-gnu/SHA256SUMS.part
6e43e1faf23c13460a779aff681c0b39ab489cdc7ffc08656492638f0d1e9745  guix-build-85888b5210c4/output/aarch64-linux-gnu/bitcoin-85888b5210c4-aarch64-linux-gnu-debug.tar.gz
a23d0c0734f80e6faa15da864a71670131ab4ca0842a866585137f253d4aa6d7  guix-build-85888b5210c4/output/aarch64-linux-gnu/bitcoin-85888b5210c4-aarch64-linux-gnu.tar.gz
3d8f004f946919d0970ab7a01b8fc8d2708f67297e2e106009428b0512c13787  guix-build-85888b5210c4/output/arm-linux-gnueabihf/SHA256SUMS.part
bd6bc1c40cf664d0044ba21b4d026ba4b43d4b14b70552fc0ddbf3a93689c519  guix-build-85888b5210c4/output/arm-linux-gnueabihf/bitcoin-85888b5210c4-arm-linux-gnueabihf-debug.tar.gz
b0cb8eede06b51a359d7d8fed7adf4b1397d814b68d2e423470df62ba1fcca83  guix-build-85888b5210c4/output/arm-linux-gnueabihf/bitcoin-85888b5210c4-arm-linux-gnueabihf.tar.gz
b41950cc73a0130989c30093a27982ebfabb9f8de31f9310f06fa6e14644eeea  guix-build-85888b5210c4/output/dist-archive/bitcoin-85888b5210c4.tar.gz
57d07de9fe2e43a1b5cc508e13a517f7d224085f19cb169627b97d4e9ba61f18  guix-build-85888b5210c4/output/powerpc64-linux-gnu/SHA256SUMS.part
07f69a38755900293550aab3c603ed46193e6f3737e0dce3419bbe7e9d96a226  guix-build-85888b5210c4/output/powerpc64-linux-gnu/bitcoin-85888b5210c4-powerpc64-linux-gnu-debug.tar.gz
9469b2c83c5df17717ddab52015a352256fcf3241bd500321438558ceec8d21c  guix-build-85888b5210c4/output/powerpc64-linux-gnu/bitcoin-85888b5210c4-powerpc64-linux-gnu.tar.gz
57c2a2384e4e0feeaf5c261e0ba05652253e10bb84c1597e8b18c0b280ed0d6d  guix-build-85888b5210c4/output/powerpc64le-linux-gnu/SHA256SUMS.part
a3587852be8dba91dac10a91f8827570677cae215afb61865dfa962b4d92044b  guix-build-85888b5210c4/output/powerpc64le-linux-gnu/bitcoin-85888b5210c4-powerpc64le-linux-gnu-debug.tar.gz
b7a2b4e450a76bae79da8012b1b9ef70467d3760d61bc8a0d8b41271249caf44  guix-build-85888b5210c4/output/powerpc64le-linux-gnu/bitcoin-85888b5210c4-powerpc64le-linux-gnu.tar.gz
cefaeb09ff6f6ed57850a1762090b58d118515d0e953d1d5eea1ad031f73dafa  guix-build-85888b5210c4/output/riscv64-linux-gnu/SHA256SUMS.part
b6897612ddca28534ffa61fd2bb321eafabcb01f3e16140934a732086930b851  guix-build-85888b5210c4/output/riscv64-linux-gnu/bitcoin-85888b5210c4-riscv64-linux-gnu-debug.tar.gz
90a4c28c2c85c562a0ba50cc158867d4634b6fa961fa61ec319fb87318f1780b  guix-build-85888b5210c4/output/riscv64-linux-gnu/bitcoin-85888b5210c4-riscv64-linux-gnu.tar.gz
ad48cd61446fbfc2e982aa2f6facced8b8c63c2606912fafea1eceadcf0ef1bd  guix-build-85888b5210c4/output/x86_64-apple-darwin18/SHA256SUMS.part
b3f866ed75e3e3eff5d6c565645abf7e6083de99c4c745d40c3053c60cebf7e5  guix-build-85888b5210c4/output/x86_64-apple-darwin18/bitcoin-85888b5210c4-osx-unsigned.dmg
c31f297161e002dd7e61bdc1ae511d6b39093c429cfe3b8f98ab51da3eca25eb  guix-build-85888b5210c4/output/x86_64-apple-darwin18/bitcoin-85888b5210c4-osx-unsigned.tar.gz
43835f47aeaa88c3b80b6623a9642f7ff6c887a41c30ab3c5f45ddafc004171b  guix-build-85888b5210c4/output/x86_64-apple-darwin18/bitcoin-85888b5210c4-osx64.tar.gz
f9a7c42bdf86aa1e16508f39f8144d3dd051a0a729fe3ffb6710e10723214ee5  guix-build-85888b5210c4/output/x86_64-linux-gnu/SHA256SUMS.part
da8b9e48cf9470e7105bd0614fefd618445cba9150a5ed3ca56b0a8f44133909  guix-build-85888b5210c4/output/x86_64-linux-gnu/bitcoin-85888b5210c4-x86_64-linux-gnu-debug.tar.gz
fbe71ec235394aed4643f5c48d4249a17b3769218757b5ac33c5a945d61c3236  guix-build-85888b5210c4/output/x86_64-linux-gnu/bitcoin-85888b5210c4-x86_64-linux-gnu.tar.gz
f10c3f187b9f19dcd0306e84bb81b06234bc48164377f9488ed457dfe024d438  guix-build-85888b5210c4/output/x86_64-w64-mingw32/SHA256SUMS.part
b59fed910a459911d7e088b5f9996ff2a5e271da1ccd3f20adc512fb64e73154  guix-build-85888b5210c4/output/x86_64-w64-mingw32/bitcoin-85888b5210c4-win-unsigned.tar.gz
b5314730b0cf1176821bbc66f9c9ecec4e97bb189733533a80f043e0db49167f  guix-build-85888b5210c4/output/x86_64-w64-mingw32/bitcoin-85888b5210c4-win64-debug.zip
472b673f781d1dd0f411ef8ee42baf40e01592c82ec8538ab7df70c9758318c5  guix-build-85888b5210c4/output/x86_64-w64-mingw32/bitcoin-85888b5210c4-win64-setup-unsigned.exe
185abcc544ea7c625be27a866e804b23885e61abbb28937ed75976996fd39668  guix-build-85888b5210c4/output/x86_64-w64-mingw32/bitcoin-85888b5210c4-win64.zip
```

fanquake commented:
I'm closing this for now. In favour of #23583.
